minio/minio
# Enable event persistence in webhook

## Description
Webhook events will be persisted in a queue dir if configured, else persisted in memory.

## Motivation and Context
To provide offline backup/persistence for events, so that no events are lost.

## Regression
It is a feature

## How Has This Been Tested?
- Start MinIO without an active nsq broker with the following config
```
"webhook": {
  "1": {
    "enable": true,
    "endpoint": "http://localhost:8080/",
    "queueDir": "/tmp/webhook",
     "queueLimit": 1000
   }
}
```
- Start sending events
- The events will persist in /tmp/webhook and gets replayed when the broker is turned on. (/tmp/webhook - will be empty if the events are successfully sent)
- Run the webhook server on port 8080
```
/*
 * Minio Cloud Storage, (C) 2019 Minio, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 package main

 import (
	 "log"
	 "net/http"
	 "io/ioutil"
	 "encoding/json"
	 "github.com/minio/minio/pkg/event"
	 "fmt"
 )


func main() {

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		b, err := ioutil.ReadAll(r.Body)
		defer r.Body.Close()
		if err != nil {
			log.Fatal(err)
			return
		}
		var msg event.Log
		err = json.Unmarshal(b, &msg)
		if err != nil {
			log.Fatal(err)
			return
		}
		fmt.Println("**********New-Message****************")
		fmt.Println(msg)
		w.WriteHeader(200)
		w.Write([]byte("ping")) 
	})

	log.Printf("listening on http://%s/", "localhost:8080")
	log.Fatal(http.ListenAndServe("localhost:8080", nil))
 }
```
- You can see the events getting replayed in the console

## Types of changes

- [] Bug fix (non-breaking change which fixes an issue)

## Checklist:
